### PR TITLE
fix(general): refactor SARIF output

### LINKED ISSUE #4585: Checkov to include affected code snippet in SARIF output
**Describe the issue**
Currently, SARIF format does **not** include affected code snippets in output files. But other output files do include code snippets.

### LINKED ISSUE #4600: Invalid URI when scanning Nginx based image
Currently we are facing an issue with scanning a Dockerfile based on Nginx (nginx:1.23.3-alpine). In this case the upload of the SARIF report fails.

https://github.com/github/codeql-action/issues/1553

Reason seems to be an invalid URI on the SARIF report file.

```
"locations": [
  {
    "physicalLocation": {
      "artifactLocation": {
        "uri": "Dockerfile (nginx:1.23.3-alpine lines:1-1 (sha256:2bc7edbc3c))"
      },
      "region": {
        "startLine": 1,
        "endLine": 1
      }
    }
  }
]
```

Any chance getting this fixed?

```
[SARIF1002](http://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html): runs[0].results[5].locations[0].physicalLocation.artifactLocation.uri: The string 'Dockerfile (nginx:1.23.3-alpine lines:1-1 (sha256:2bc7edbc3c))' is not a valid URI reference. URIs must conform to [RFC 3986](https://tools.ietf.org/html/rfc3986).
```

### PR DESCRIPTION
**Pull request by gruebel**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

- adds affected code lines to the report
- fixes an issue with IR output
- moved and refactored the SARIF code to a dedicated class
- rule IDs for licenses and CVEs are now correct and don't smooshed to a few
  - GPL_geoip@1.6.12-r3
  - CVE-2023-23914_curl@7.87.0-r1
- also added a doc page for SARIF

Fixes #4600
Fixes #4585 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by ayajbara — APPROVED** [review]

File: checkov/common/output/sarif.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+                        "organization": "bridgecrew",
+                    }
+                },
+                "results": results,
+            }
+        ]
+
+    def _create_rules(self) -> list[dict[str, Any]]:
+        rule_idx = 0
+        rules: "list[dict[str, Any]]" = []
```
Comment (Eliran-Turgeman) [inline checkov/common/output/sarif.py:61]: Is quoted typing necessary even if you imported annotations from __future__?

File: checkov/common/output/sarif.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        return None
+
+    def write_sarif_output(self) -> None:
+        try:
+            with open("results.sarif", "w") as f:
+                f.write(json.dumps(self.json))
+                print("\nWrote output in SARIF format to the file 'results.sarif'")
+        except EnvironmentError as e:
+            print("\nAn error occurred while writing SARIF results to file: results.sarif")
+            print(f"More details: \n {e}")
```
Comment (Eliran-Turgeman) [inline checkov/common/output/sarif.py:242]: shouldn't we logging instead of print here?

**Review by marynaKK — APPROVED** [review]

File: checkov/common/output/sarif.py (hunk 1 above)
Comment (gruebel) [reply checkov/common/output/sarif.py:61]: it is a bit faster to put into a string literal, I somehow got used to it, when I add complex type hints. But you are right it is not needed when using the `__future__` import

Comment (Eliran-Turgeman) [reply checkov/common/output/sarif.py:61]: I dug the rabbit hole a bit further and found out that `"list[dict[str, Any]]" != list[dict[str, Any]]`
It seems to have no real effect on linting for now but may be worth changing it regardless (?)
https://stackoverflow.com/a/46458497/11918269

Comment (gruebel) [reply checkov/common/output/sarif.py:61]: and as the comment to the answer suggests, don't do that and use the proper function `typing.get_type_hints()`
```python
>>> typing.get_type_hints(bad_foo) == typing.get_type_hints(good_foo)
True
```
😉

Comment (Eliran-Turgeman) [reply checkov/common/output/sarif.py:61]: 🤯

**Pull request merged**